milli/src/vector/parsed_vectors.rs

```diff
@@ -0,0 +1,208 @@
+#[serde(untagged)]
+pub enum Vectors {
+    ImplicitlyUserProvided(VectorOrArrayOfVectors),
+    Explicit(ExplicitVectors),
+}
+
+impl Vectors {
+    pub fn into_array_of_vectors(self) -> Vec<Embedding> {
+        match self {
+            Vectors::ImplicitlyUserProvided(embeddings)
+            | Vectors::Explicit(ExplicitVectors { embeddings, user_provided: _ }) => {
+                embeddings.into_array_of_vectors().unwrap_or_default()
+            }
+        }
+    }
+}
+
+#[derive(serde::Serialize, serde::Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct ExplicitVectors {
+    pub embeddings: VectorOrArrayOfVectors,
+    pub user_provided: bool,
+}
+
+pub struct ParsedVectorsDiff {
+    pub old: Option<BTreeMap<String, Vectors>>,
+    pub new: Option<BTreeMap<String, Vectors>>,
+}
+
+impl ParsedVectorsDiff {
+    pub fn new(
+        documents_diff: KvReader<'_, FieldId>,
+        old_vectors_fid: Option<FieldId>,
+        new_vectors_fid: Option<FieldId>,
+    ) -> Result<Self, Error> {
+        let old = match old_vectors_fid
+            .and_then(|vectors_fid| documents_diff.get(vectors_fid))
+            .map(KvReaderDelAdd::new)
+            .map(|obkv| to_vector_map(obkv, DelAdd::Deletion))
+            .transpose()
+        {
+            Ok(del) => del,
+            // ignore wrong shape for old version of documents, use an empty map in this case
+            Err(Error::InvalidMap(value)) => {
+                tracing::warn!(%value, "Previous version of the `_vectors` field had a wrong shape");
+                Default::default()
+            }
+            Err(error) => {
+                return Err(error);
+            }
+        }
+        .flatten();
+        let new = new_vectors_fid
+            .and_then(|vectors_fid| documents_diff.get(vectors_fid))
+            .map(KvReaderDelAdd::new)
+            .map(|obkv| to_vector_map(obkv, DelAdd::Addition))
+            .transpose()?
+            .flatten();
+        Ok(Self { old, new })
+    }
+
+    pub fn remove(&mut self, embedder_name: &str) -> (Option<Vectors>, Option<Vectors>) {
+        let old = self.old.as_mut().and_then(|old| old.remove(embedder_name));
+        let new = self.new.as_mut().and_then(|new| new.remove(embedder_name));
+        (old, new)
+    }
+}
+
+pub struct ParsedVectors(pub BTreeMap<String, Vectors>);
+
+impl ParsedVectors {
+    pub fn from_bytes(value: &[u8]) -> Result<Self, Error> {
+        let Ok(value) = from_slice(value) else {
+            let value = from_slice(value).map_err(Error::InternalSerdeJson)?;
+            return Err(Error::InvalidMap(value));
+        };
+        Ok(ParsedVectors(value))
+    }
+
+    pub fn retain_user_provided_vectors(&mut self) {
+        self.0.retain(|_k, v| match v {
+            Vectors::ImplicitlyUserProvided(_) => true,
+            Vectors::Explicit(ExplicitVectors { embeddings: _, user_provided }) => *user_provided,
+        });
+    }
```

Comment:
This function should keep vectors associated with an embedder that does not exist.

Reply:
Changed to do so